src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts

```diff
@@ -48,6 +52,32 @@ export class WorkspaceIdConsumerWrapper {
     return type === UI_SETTINGS_SAVED_OBJECTS_TYPE;
   }
 
+  private validateObjectInAWorkspace<T>(
+    object: SavedObject<T>,
+    workspace: string,
+    request: OpenSearchDashboardsRequest
+  ) {
+    // Data source is a workspace level object, validate if the request has access to the data source within the requested workspace.
+    if (validateIsWorkspaceDataSourceAndConnectionObjectType(object.type)) {
+      if (!!getWorkspaceState(request).isDataSourceAdmin) {
+        return true;
+      }
+      // Deny access if the object is a global data source (no workspaces assigned)
+      if (!object.workspaces || object.workspaces.length === 0) {
+        return false;
+      }
+    }
+    /*
+     * Allow access if the requested workspace matches one of the object's assigned workspaces
+     * This ensures that the user can only access data sources within their current workspace
+     */
+    if (object.workspaces && object.workspaces.length > 0) {
+      return object.workspaces.includes(workspace);
+    }
+    // Allow access if the object is a global object (object.workspaces is null/[])
```

Comment:
The comment looks weird to me, if the object is global, they will be denied in line 66.